ape_safe/factory.py

```diff
@@ -0,0 +1,115 @@
+import secrets
+from functools import cache
+from typing import TYPE_CHECKING, Iterable, Optional, Union
+
+from ape.types import AddressType
+from ape.utils import ZERO_ADDRESS, ManagerAccessMixin
+from packaging.version import Version
+
+from ape_safe.packages import SAFE_PACKAGE_BY_VERSION, PackageType, get_factory, get_singleton
+
+if TYPE_CHECKING:
+    from ape.api import AccountAPI
+    from ape.contracts import ContractInstance
+
+
+class Factory(ManagerAccessMixin):
+    _singleton: dict[Version, "ContractInstance"] = {}
+    _factory: dict[Version, "ContractInstance"] = {}
+
+    @classmethod
+    def inject(cls, version: Version, deployer: "AccountAPI"):
+        cls._singleton[version] = deployer.deploy(PackageType.SINGLETON(version))
+        cls._factory[version] = deployer.deploy(PackageType.PROXY_FACTORY(version))
+
+    @cache
+    def get_factory(self, version: Version) -> "ContractInstance":
+        if injected_factory := self._factory.get(version):
+            return injected_factory
+
+        return get_factory(self.chain_manager.chain_id, version)
+
+    @property
+    def contract(self) -> "ContractInstance":
+        return self.get_factory(max(SAFE_PACKAGE_BY_VERSION))
+
+    @cache
+    def get_singleton(self, version: Version) -> "ContractInstance":
+        if injected_singleton := self._singleton.get(version):
+            return injected_singleton
+
+        return get_singleton(self.chain_manager.chain_id, version)
+
+    def create(
+        self,
+        owners: Iterable[Union["AddressType", str]],
```

Comment:
It seems like `BaseAddress` would also work here

Reply:
3814c70